File: src/entities/Order.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToOne,
  JoinColumn,
  OneToMany,
  CreateDateColumn,
  UpdateDateColumn,
} from "typeorm";
import { User } from "./User";
import { Restaurant } from "./Restaurant";
import { OrderItem } from "./OrderItem";

export enum OrderStatus {
  CRIADO = "CRIADO",
  ACEITO = "ACEITO",
  ENTREGADOR_DESIGNADO = "ENTREGADOR_DESIGNADO",
  EM_ENTREGA = "EM_ENTREGA",
  ENTREGUE = "ENTREGUE",
}

const VALID_TRANSITIONS: Record<OrderStatus, OrderStatus | null> = {
  [OrderStatus.CRIADO]: OrderStatus.ACEITO,
  [OrderStatus.ACEITO]: OrderStatus.ENTREGADOR_DESIGNADO,
  [OrderStatus.ENTREGADOR_DESIGNADO]: OrderStatus.EM_ENTREGA,
  [OrderStatus.EM_ENTREGA]: OrderStatus.ENTREGUE,
  [OrderStatus.ENTREGUE]: null,
};

export function isValidTransition(from: OrderStatus, to: OrderStatus): boolean {
  return VALID_TRANSITIONS[from] === to;
}

@Entity("orders")
export class Order {
  @PrimaryGeneratedColumn("uuid")
  id!: string;

  @Column()
  clientId!: string;

  @ManyToOne(() => User)
  @JoinColumn({ name: "clientId" })
  client!: User;

  @Column()
  restaurantId!: string;

  @ManyToOne(() => Restaurant, (restaurant) => restaurant.orders)
  @JoinColumn({ name: "restaurantId" })
  restaurant!: Restaurant;

  @Column({ nullable: true })
  deliveryPersonId?: string;

  @ManyToOne(() => User, { nullable: true })
  @JoinColumn({ name: "deliveryPersonId" })
  deliveryPerson?: User;

  @Column({ type: "enum", enum: OrderStatus, default: OrderStatus.CRIADO })
  status!: OrderStatus;

  @Column({ type: "decimal", precision: 10, scale: 2 })
  total!: number;

  @Column()
  deliveryAddress!: string;

  @OneToMany(() => OrderItem, (orderItem) => orderItem.order, { cascade: true })
  items!: OrderItem[];

  @CreateDateColumn()
  createdAt!: Date;

  @UpdateDateColumn()
  updatedAt!: Date;
}

File: flutter_app/lib/core/models/order.dart
import 'order_item.dart';

enum OrderStatus {
  CRIADO,
  ACEITO,
  ENTREGADOR_DESIGNADO,
  EM_ENTREGA,
  ENTREGUE;

  String get label => switch (this) {
    CRIADO => 'Recebido',
    ACEITO => 'Aceito',
    ENTREGADOR_DESIGNADO => 'Entregador a caminho',
    EM_ENTREGA => 'Em entrega',
    ENTREGUE => 'Entregue',
  };
}

class Order {
  final String id;
  final String clientId;
  final String restaurantId;
  final String? deliveryPersonId;
  final OrderStatus status;
  final double total;
  final String deliveryAddress;
  final List<OrderItem> items;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Order({
    required this.id,
    required this.clientId,
    required this.restaurantId,
    this.deliveryPersonId,
    required this.status,
    required this.total,
    required this.deliveryAddress,
    this.items = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) => Order(
    id: json['id'] as String,
    clientId: json['clientId'] as String,
    restaurantId: json['restaurantId'] as String,
    deliveryPersonId: json['deliveryPersonId'] as String?,
    status: OrderStatus.values.firstWhere((s) => s.name == json['status']),
    total: (json['total'] as num).toDouble(),
    deliveryAddress: json['deliveryAddress'] as String,
    items: (json['items'] as List<dynamic>?)
        ?.map((e) => OrderItem.fromJson(e as Map<String, dynamic>))
        .toList() ?? [],
    createdAt: DateTime.parse(json['createdAt'] as String),
    updatedAt: DateTime.parse(json['updatedAt'] as String),
  );

  Order copyWith({OrderStatus? status, DateTime? updatedAt}) => Order(
    id: id,
    clientId: clientId,
    restaurantId: restaurantId,
    deliveryPersonId: deliveryPersonId,
    status: status ?? this.status,
    total: total,
    deliveryAddress: deliveryAddress,
    items: items,
    createdAt: createdAt,
    updatedAt: updatedAt ?? this.updatedAt,
  );
}
